rubix-core/src/main/java/com/qubole/rubix/core/CachingFileSystem.java

```diff
@@ -99,26 +99,36 @@ private static void initialize(Configuration conf, ClusterType clusterType)
 
   private static void initializeStats(Configuration conf)
   {
-    MBeanExporter exporter = new MBeanExporter(ManagementFactory.getPlatformMBeanServer());
-    stats = new CachingFileSystemStatsProvider();
-    try {
-      if (ManagementFactory.getPlatformMBeanServer().isRegistered(new ObjectName(statsMBeanName))) {
-        exporter.unexport(statsMBeanName);
-      }
-      if (!ManagementFactory.getPlatformMBeanServer().isRegistered(new ObjectName(statsMBeanName))) {
-        exporter.export(statsMBeanName, new BasicCachingFileSystemStats(stats, bookKeeperFactory, conf));
-      }
+    if (statsInitialized.get()) {
+      return;
+    }
 
-      if (ManagementFactory.getPlatformMBeanServer().isRegistered(new ObjectName(detailedStatsMBeanName))) {
-        exporter.unexport(detailedStatsMBeanName);
+    synchronized (statsInitialized) {
+      if (statsInitialized.get()) {
+        return;
       }
-      if (!ManagementFactory.getPlatformMBeanServer().isRegistered(new ObjectName(detailedStatsMBeanName))) {
-        exporter.export(detailedStatsMBeanName, new DetailedCachingFileSystemStats(stats, bookKeeperFactory, conf));
+
+      MBeanExporter exporter = new MBeanExporter(ManagementFactory.getPlatformMBeanServer());
+      stats = new CachingFileSystemStatsProvider();
+      try {
+        if (ManagementFactory.getPlatformMBeanServer().isRegistered(new ObjectName(statsMBeanName))) {
```

Comment:
if the initialization is idempotent, it doesn't make sense to unregister stats

Reply:
It does with current changes but given that we are removing the `checkState(!initialized.get()` from `setLocalBookKeeper` => we are allowing re-init of stats so we will need to keep this.